mirrord/cli/src/port_forward.rs

```diff
@@ -442,12 +420,255 @@ impl PortForwarder {
+            }
+        }
+    }
+
+    #[tracing::instrument(level = Level::TRACE, skip(self), err)]
+    async fn handle_msg_from_agent(
+        &mut self,
+        message: DaemonMessage,
+    ) -> Result<(), PortForwardError> {
+        match message {
+            DaemonMessage::Tcp(msg) => {
+                self.incoming_proxy
+                    .send(IncomingProxyMessage::AgentMirror(msg))
+                    .await
+            }
+            DaemonMessage::TcpSteal(msg) => {
+                self.incoming_proxy
+                    .send(IncomingProxyMessage::AgentSteal(msg))
+                    .await
+            }
+            DaemonMessage::LogMessage(log_message) => match log_message.level {
+                LogLevel::Warn => tracing::warn!("agent log: {}", log_message.message),
+                LogLevel::Error => tracing::error!("agent log: {}", log_message.message),
+            },
+            DaemonMessage::Close(error) => {
+                return Err(PortForwardError::AgentError(error));
+            }
+            DaemonMessage::Pong if self.waiting_for_pong => {
+                self.waiting_for_pong = false;
+            }
+            other => {
+                // includes unexepcted DaemonMessage::Pong
+                return Err(PortForwardError::AgentError(format!(
+                    "unexpected message from agent: {other:?}"
+                )));
+            }
+        }
+
+        Ok(())
+    }
+
+    #[tracing::instrument(level = Level::TRACE, skip(self), err)]
+    async fn handle_msg_from_local(
+        &mut self,
+        task_id: MainTaskId,
+        update: TaskUpdate<ProxyMessage, IntProxyError>,
+    ) -> Result<(), PortForwardError> {
+        match (task_id, update) {
+            (MainTaskId::IncomingProxy, TaskUpdate::Message(message)) => match message {
+                ProxyMessage::ToAgent(message) => {
+                    if matches!(
+                        message,
+                        ClientMessage::TcpSteal(LayerTcpSteal::PortSubscribe(_))
+                    ) || matches!(message, ClientMessage::Tcp(LayerTcp::PortSubscribe(_)))
+                    {
+                        // suppress additional subscription requests
+                        return Ok(());
+                    }
+                    self.agent_connection.sender.send(message).await?;
+                }
+                ProxyMessage::ToLayer(ToLayer { message, .. }) => match message {
+                    ProxyToLayerMessage::Incoming(IncomingResponse::PortSubscribe(res)) => {
+                        if let Err(error) = res {
+                            return Err(PortForwardError::from(IntProxyError::from(
+                                IncomingProxyError::SubscriptionFailed(error),
+                            )));
+                        }
+                    }
+                    _ => {
+                        unreachable!("message type are never used in port forwarding: {message:?}")
```

Comment:
```suggestion
                        unreachable!("message type is never used in port forwarding: {message:?}")
```

right?

Reply:
Thinking about it, maybe we should not panic if this happen, and instead just log and then ignore the message.
I'm thinking about what if we add new message types in the future, that can originate in the agent without the client sending a request first. WDYT?

Reply:
yep, sounds sensible 👍